attrd/attrd_alerts.c

```diff
@@ -0,0 +1,566 @@
+                entry->select_attribute_name_orig, attribute_name);
+            free(timestamp);
+            continue;
+        }
+
+        crm_info("Sending '%s' alert to '%s' via '%s'", kind, entry->recipient, entry->path);
+
+        rsc = g_hash_table_lookup(alert_info_cache, entry->id);
+        if (rsc == NULL) {
+            rsc = lrmd->cmds->get_rsc_info(lrmd, entry->id, 0);
+            if (!rsc) {
+                lrmd->cmds->register_rsc(lrmd, entry->id, PCMK_ALERT_CLASS,  "pacemaker", entry->path, lrmd_opt_drop_recurring);
+                rsc = lrmd->cmds->get_rsc_info(lrmd, entry->id, 0);
+                if (!rsc) {
+                    crm_err("Could not add alert %s : %s", entry->id, entry->path);
+                    return -1; 
+                }
+                /* cache the result */
+                g_hash_table_insert(alert_info_cache, entry->id, rsc);
+            }
+        }
+
+        /* Because there is a parameter to turn into every transmission, Copy a parameter. */
+        head = params;
+        while (head) {
+            p = head->next;
+            copy_params = lrmd_key_value_add(copy_params, head->key, head->value);
+            head = p;
+        }
+
+        operations++;
+
+        copy_params = lrmd_key_value_add(copy_params, CRM_ALERT_KEY_PATH, entry->path);
+        copy_params = lrmd_set_alert_key_to_lrmd_params(copy_params, CRM_alert_recipient, entry->recipient);
+        copy_params = lrmd_set_alert_key_to_lrmd_params(copy_params, CRM_alert_node_sequence, crm_itoa(operations));
+        copy_params = lrmd_set_alert_key_to_lrmd_params(copy_params, CRM_alert_timestamp, timestamp);
+
+        lrmd_set_alert_envvar_to_lrmd_params(entry, copy_params);
+        
+        call_id = lrmd->cmds->exec_alert(lrmd, strdup(entry->id), entry->timeout, lrmd_opt_notify_orig_only, copy_params);
+        if (call_id <= 0) {
+            crm_err("Operation %s on %s failed: %d", "start", rsc->id, call_id);
+        } else {
+            crm_info("Operation %s on %s compete: %d", "start", rsc->id, call_id);
+        }
+
+        free(timestamp);
+    }
+
+    if (now) {
+        free(now);
+    }
+
+    return call_id;
+}
+
+static void
+free_alert_info(gpointer value)
+{
+    lrmd_rsc_info_t *rsc_info = value;
+
+    lrmd_free_rsc_info(rsc_info);
+}
+
+static void
+attrd_alert_lrm_op_callback(lrmd_event_data_t * op)
+{
+    const char *nodename = NULL;
+
+    CRM_CHECK(op != NULL, return);
+
+#if HAVE_ATOMIC_ATTRD
+    nodename = op->remote_nodename ? op->remote_nodename : attrd_cluster->uname;
+#else
+    nodename = op->remote_nodename ? op->remote_nodename : attrd_uname;
+#endif
+
+#if HAVE_ATOMIC_ATTRD
+    if (op->type == lrmd_event_disconnect && (safe_str_eq(nodename, attrd_cluster->uname))) {
+        crm_info("Lost connection to LRMD service!");
+#else
+    if (op->type == lrmd_event_disconnect && (safe_str_eq(nodename, attrd_uname))) {
+        crm_notice("Lost connection to LRMD service!");
+#endif
```

Comment:
I intend to make difference in next degree.
What do you think?

```
static void
attrd_alert_lrm_op_callback(lrmd_event_data_t * op)
{
    const char *nodename = NULL;
    const char *local_attrd_uname = NULL;
    CRM_CHECK(op != NULL, return);

#if HAVE_ATOMIC_ATTRD
    nodename = op->remote_nodename ? op->remote_nodename : attrd_cluster->uname;
    local_attrd_uname = attrd_cluster->uname;
#else
    nodename = op->remote_nodename ? op->remote_nodename : attrd_uname;
    local_attrd_uname = attrd_uname;
#endif
    if (op->type == lrmd_event_disconnect && (safe_str_eq(nodename, local_attrd_uname))) {        crm_info("Lost connection to LRMD service!");        if (the_lrmd->cmds->is_connected(the_lrmd)) {            the_lrmd->cmds->disconnect(the_lrmd);            lrmd_api_delete(the_lrmd);        }        the_lrmd = NULL;
        return;
    } else if (op->type != lrmd_event_exec_complete) {
        return;
    }

    if (op->params != NULL) {
        void *value_tmp1, *value_tmp2;
 
       value_tmp1 = g_hash_table_lookup(op->params, CRM_ALERT_KEY_PATH);
        if (value_tmp1 != NULL) {
            value_tmp2 = g_hash_table_lookup(op->params, CRM_ALERT_NODE_SEQUENCE);
            if(op->rc == 0) {
#if HAVE_ATOMIC_ATTRD
                crm_info("Alert %s (%s) complete", value_tmp2, value_tmp1);
#else
                crm_notice("Alert %s (%s) complete", value_tmp2, value_tmp1);
#endif
            } else {
                crm_warn("Alert %s (%s) failed: %d", value_tmp2, value_tmp1, op->rc);
            }
        }
    }
}
```
Is there still unnecessary difference?